docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,995 @@
+users who want to specify their endorsements (information from a second source
+used for verification), instead of using the set of endorsements provided by Open
+Enclave.
+
+Overall, there has been interest in enhancing Open Enclave's APIs to support
+custom attestation formats to enable these scenarios.
+
+Terminology
+-----------
+
+This document uses the following terminology defined below. Note that
+these definitions are consistent with the terms defined in the
+[Remote Attestation Procedures (RATS)](https://datatracker.ietf.org/wg/rats/about/)
+working group.
+
+- Claims
+  - Claims are statements about a particular subject. They consist of
+    name-value pairs containing the claim name, which is a string, and
+    claim value, which is arbitrary data. Example of claims could be
+    [name="version", value=1] or [name="enclave_id", value=1111].
+- Evidence
+  - Evidence are claims about the enclave that are produced and signed by it.
+    The SGX report would be an example of evidence.
+- Endorsements
+  - Endorsements are additional claims used in the evidence verification process,
+    but are not produced by the enclave. An example of an endorsement would be
+    the quoting enclave's identity used in SGX remote attestation, because it
+    is retrieved from Intel's servers, rather than the enclave.
+- Attester
+  - The attester creates the evidence and signs it. Trusted Execution Environments
+    (TEEs), such as the SGX enclave, often play the role of the attester.
+- Verifier
+  - The verifier is responsible for taking in the evidence and endorsements
+    and deciding if the enclave is trustworthy.
+- Relying party
+  - The relying party is the entity interested in communicating with an
+    enclave. The enclave must attest to the relying party before the
+    relying party can trust it. The relying party can also play the role
+    of the verifier, but it does not necessarily have to.
+
+Specification
+-------------
+
+To support custom attestation formats, this document proposes adding a plugin
+model for attestation. The Open Enclave SDK will define a plugin API for the
+attester and another API for the verifier. Each plugin will define a UUID to
+distinguish plugins. An attester and verifier plugin sharing the same UUID
+indicates that that verifier is able to process the evidence format generated
+by the attester.
+
+Futhermore, there will be additional attestation "plugin aware" APIs that are
+analogous to `oe_get_report` and `oe_verify_report` called `oe_get_evidence`
+and `oe_verify_evidence` respectively. There will also
+be functions for registering and unregistering plugins called
+`oe_register_[attester|verifier]` and `oe_unregister_[attester|verfier]`. The user
+can link in their desired plugin and call the register plugin function.
+The attestation data can be retrieved from `oe_get_evidence` with the desired UUID.
+The generated data will have the UUID in its header. The user can `oe_verify_evidence`
```

Comment:
This line refers to the "UUID in its header".   Is this referring to some blob that needs to be passed over a network?  Or is it just internal to a machine?

If it needs to be passed over a network, then it presumably would be better to have a standard there rather than some OE-specific format, yes?   Would like to understand so I can help in bringing an ask to IETF or whatever other body if needed.

Reply:
Yeah it will need to be passed over the network. The verifier gets the blob from the network and then calls `oe_verify_evidence`, which then parses the blob header to determine what verifier plugin to run using the UUID.

 It would be better if it's in a standard, but I didn't see something equivalent in the RATS standard. There's an UUID to uniquely identify the device, but I didn't see anything to distinguish different data formats in the paper.